components/web_discovery/browser/url_extractor.h

```diff
@@ -0,0 +1,91 @@
+/* Copyright (c) 2025 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_WEB_DISCOVERY_BROWSER_URL_EXTRACTOR_H_
+#define BRAVE_COMPONENTS_WEB_DISCOVERY_BROWSER_URL_EXTRACTOR_H_
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "base/memory/raw_ptr.h"
+#include "brave/components/web_discovery/browser/relevant_site.h"
+#include "third_party/re2/src/re2/re2.h"
+#include "url/gurl.h"
+
+namespace web_discovery {
+
+// Details about a relevant site pattern
+struct RelevantSiteDetails {
+  RelevantSiteDetails(RelevantSite site,
+                      bool is_search_engine,
+                      std::unique_ptr<re2::RE2>&& regex,
+                      std::vector<std::string_view>&& query_params,
+                      std::optional<std::string_view> prefix);
+  ~RelevantSiteDetails();
+
+  RelevantSiteDetails(const RelevantSiteDetails&) = delete;
+  RelevantSiteDetails& operator=(const RelevantSiteDetails&) = delete;
+
+  RelevantSiteDetails(RelevantSiteDetails&&);
+  RelevantSiteDetails& operator=(RelevantSiteDetails&&);
+
+  RelevantSite site;
+  bool is_search_engine;
+  std::unique_ptr<re2::RE2> regex;
+  // List of query parameters to process in order to extract the query
+  std::vector<std::string_view> query_params;
+  // Used for generating a private search URL
+  std::optional<std::string_view> private_query_prefix;
+};
+
+// Result of URL identification
+struct URLExtractResult {
+  URLExtractResult(const RelevantSiteDetails* details,
+                   std::optional<std::string>&& query);
+  ~URLExtractResult();
+
+  URLExtractResult(const URLExtractResult&) = delete;
+  URLExtractResult& operator=(const URLExtractResult&) = delete;
+
+  URLExtractResult(URLExtractResult&&);
+  URLExtractResult& operator=(URLExtractResult&&);
+
+  raw_ptr<const RelevantSiteDetails> details;
+  std::optional<std::string> query;
+};
+
+// URLExtractor provides functionality for identifying URLs and extracting
```

Comment:
don't we already have this functionality in `TemplateURLService::ExtractSearchMetadata`?

Reply:
There is some overlap with `TemplateURL`, but I decided to create this lightweight identifier/extractor because WDP uses regex matching to identify multiple hosts for a given site ID (i.e. country-specific search engines), and I wanted to ensure that the list of regexes and functionality is comparable to the reference implementation.